cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package upgrade
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+)
+
+//go:embed lua/upgrade.lua
+var upgradeLuaScript string
+
+type SyncInfo struct {
+	LSN uint32  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+func internalUpgrade(conn connector.Connector) (SyncInfo, error) {
+	var upgradeInfo SyncInfo
+	res, err := conn.Eval(upgradeLuaScript, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return upgradeInfo, err
+	}
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return upgradeInfo, err
+	}
+	if upgradeInfo.Err != nil {
+		return upgradeInfo, errors.New(*upgradeInfo.Err)
+	}
+	return upgradeInfo, nil
+}
+
+func WaitLSN(conn connector.Connector, masterUpgradeInfo SyncInfo, timeout int) error {
+	for i := 0; i < timeout; i++ {
+		var res []interface{}
+		res, err := conn.Eval(
+			fmt.Sprintf("return box.info.vclock[%d]", masterUpgradeInfo.IID),
+			[]any{}, connector.RequestOpts{})
+		if err != nil || len(res) == 0 {
+			return err
+		}
+		var lsn uint32
+		switch v := res[0].(type) {
+		case uint16:
+			lsn = uint32(v)
+		case uint32:
+			lsn = v
+		}
+		if lsn >= masterUpgradeInfo.LSN {
+			return nil
+		}
+		time.Sleep(time.Second)
+	}
+	return errors.New("")
+}
+
+func GetAllowedReplicasets(allReplicasets replicaset.Replicasets,
+	pendingReplicasetAliases []string) ([]replicaset.Replicaset, error) {
+	if len(pendingReplicasetAliases) == 0 {
+		return allReplicasets.Replicasets, nil
+	}
+
+	replicasetMap := make(map[string]replicaset.Replicaset)
+	for _, rs := range allReplicasets.Replicasets {
+		replicasetMap[rs.Alias] = rs
+	}
+
+	var allowedReplicasets []replicaset.Replicaset
+
+	for _, alias := range pendingReplicasetAliases {
+		replicaset, exists := replicasetMap[alias]
+		if !exists {
+			return nil, fmt.Errorf("replicaset with alias %q doesn't exist", alias)
+		}
+		allowedReplicasets = append(allowedReplicasets, replicaset)
+	}
+
+	return allowedReplicasets, nil
+}
+
+func Upgrade(replicasets replicaset.Replicasets, pendingReplicasetAliases []string,
+	lsnTimeout int) error {
+	allowedReplicasets, err := GetAllowedReplicasets(replicasets, pendingReplicasetAliases)
+	if err != nil {
+		return err
+	}
+
+	printReplicasetStatus := func(alias, status string) {
+		fmt.Printf("• %s: %s\n", alias, status)
+	}
+
+	for _, rs := range allowedReplicasets {
```

Comment:
I'm a bit surpsired that `tt` has no utility function that collects RO/RW information for a replicaset/a cluster. @psergee is it really so?

Anyway, I would comment this loop with something like 'collect RO/RW information', this way it is easier to spot an idea of this block on a glance. Or, better, exclude it isnot a function.

Reply:
> utility function that collects RO/RW

I found a functionality to check RO/RW.

 It now operates in two stages: 
 - if `instance.Mode == 'unknown'`, a network call is made to determine the mode;
 - otherwise, the value is taken directly from `instance.Mode`.

Reply:
Thanks, my initial point is solved. I'll look into the precise algorithm how RO/RW is determined and will open separate comments if there is something to discuss.